polygeist/tools/cgeist/Lib/Attributes.h

```diff
@@ -9,90 +9,216 @@
+/// \class
+/// This class holds the attributes for a function, its return value, and
+/// its parameters. You access the attributes for each of them via an index into
+/// the AttributeList object. The function attributes are at index
+/// `AttributeList::FunctionIndex', the return value is at index
+/// `AttributeList::ReturnIndex', and the attributes for the parameters start at
+/// index `AttributeList::FirstArgIndex'.
+class AttributeList {
+public:
+  AttributeList() = default;
+
+  //===--------------------------------------------------------------------===//
+  // AttributeList Mutation
+  //===--------------------------------------------------------------------===//
+
+  /// Add function. return value, and parameters attributes to the list.
+  AttributeList &addAttrs(const AttrBuilder &FnAttrB,
+                          const AttrBuilder &RetAttrB,
+                          llvm::ArrayRef<mlir::NamedAttrList> Attrs);
+
+  /// Add function attributes to the list.
+  AttributeList &addFnAttrs(const AttrBuilder &B);
+
+  /// Add return value attributes to the list.
+  AttributeList &addRetAttrs(const AttrBuilder &B);
+
+  /// Add parameters attributes to the list.
+  AttributeList &addParmAttrs(llvm::ArrayRef<mlir::NamedAttrList> Attrs);
+
+  /// The function attributes are returned.
+  mlir::NamedAttrList getFnAttrs() const { return FnAttrs; }
+
+  /// The attributes for the ret value are returned.
+  mlir::NamedAttrList getRetAttrs() const { return RetAttrs; }
+
+  /// The attributes for the parameters are returned.
+  mlir::ArrayRef<mlir::NamedAttrList> getParmAttrs() const { return ParmAttrs; }
+
+  /// The attributes for the parameter at the given index are returned.
+  mlir::NamedAttrList getParmAttrs(unsigned Index) const;
+
+private:
+  /// The attributes that we are managing.
+  mlir::NamedAttrList FnAttrs;
+  mlir::NamedAttrList RetAttrs;
+  llvm::SmallVector<mlir::NamedAttrList, 8> ParmAttrs;
+};
+
 /// \class
 /// Facilitates the construction of LLVM dialect attributes for a particular
 /// argument, parameter, function, or return value.
 class AttrBuilder {
 public:
-  AttrBuilder(mlir::MLIRContext &ctx) : ctx(ctx) {}
+  AttrBuilder(mlir::MLIRContext &Ctx) : Ctx(Ctx) {}
   AttrBuilder(const AttrBuilder &) = delete;
   AttrBuilder(AttrBuilder &&) = default;
 
-  /// Add the LLVM attribute identified by \p kind to the builder.
-  /// Note: only unit attributes can be added using this member function.
-  AttrBuilder &addAttribute(llvm::Attribute::AttrKind kind);
-
-  /// Add the LLVM attribute identified by \p kind with a type given by \p Ty
-  /// to the builder.
-  AttrBuilder &addAttribute(llvm::Attribute::AttrKind kind, mlir::Type Ty);
-
-  /// Add the LLVM attribute identified by \p kind with a value given by \p val
+  using AddAttrFuncPtr =
+      AttrBuilder &(AttrBuilder::*)(mlir::NamedAttribute Attr);
+  using AddRawIntAttrFuncPtr = AttrBuilder &(
+      AttrBuilder::*)(llvm::Attribute::AttrKind Kind, uint64_t Value);
```

Comment:
Can we remove this and keep the `addAttribute`/`addPassThroughAttribute` interface for clarity? In which cases will the user use this over the other interface?

Reply:
The `addAttribute`/`addPassThroughAttribute` interface is still present. So the user of these interface can continue to use those functions unchanged. The implementation details have changed to avoid duplicating code, in principle both functions add attributes (either as 'passthrough' or not) so it made sense to avoid duplicating code and trying to make the 'passthrough' member function use the other functions.

Reply:
Yes, I meant that this makes total sense in the implementation, as it makes the code way cleaner and reduces duplication, but I'd drop this from the interface and keep it in the implementation.

Reply:
Good suggestion. Will do.